java/src/test/java/ai/rapids/cudf/TableTest.java

```diff
@@ -2219,6 +2221,79 @@ void testGroupBySum() {
     }
   }
 
+  @Test
+  void testGroupByFirstExcludeNulls() {
+    try (Table t1 = new Table.TestBuilder()
+            .column(  1,   1,    1,  1,  2,    2,  2,    2)
+            .column(null, 13, null, 12, 14, null, 15, null)
+            .build()) {
+
+      try(Table t2 = t1.groupBy(0)
+              .aggregate(
+                      first(1, false));
+          HostColumnVector firstExcludeNulls = t2.getColumn(1).copyToHost()) {
+
+        assertEquals(2, firstExcludeNulls.getRowCount());
+        assertEquals(13, firstExcludeNulls.getInt(0));
+        assertEquals(14, firstExcludeNulls.getInt(1));
+      }
+    }
+  }
+
+  @Test
+  void testGroupByLastExcludeNulls() {
+    try (Table t1 = new Table.TestBuilder()
+            .column(  1,   1,    1,  1,  2,    2,  2,    2)
+            .column(null, 13, null, 12, 14, null, 15, null)
+            .build()) {
+
+      try(Table t2 = t1.groupBy(0)
+              .aggregate(last(1, false));
+          HostColumnVector lastExcludeNulls = t2.getColumn(1).copyToHost()) {
+
+        assertEquals(2, lastExcludeNulls.getRowCount());
+        assertEquals(12, lastExcludeNulls.getInt(0));
+        assertEquals(15, lastExcludeNulls.getInt(1));
+      }
+    }
+  }
+
+  @Test
+  void testGroupByFirstIncludeNulls() {
+    try (Table t1 = new Table.TestBuilder()
+            .column(  1,   1,    1,  1,  2,    2,  2,    2)
+            .column(null, 13, null, 12, 14, null, 15, null)
+            .build()) {
+
+      try(Table t2 = t1.groupBy(0)
+              .aggregate(first(1, true));
+          HostColumnVector firstIncludeNulls = t2.getColumn(1).copyToHost()) {
+
+        assertEquals(2, firstIncludeNulls.getRowCount());
+        assert(firstIncludeNulls.isNull(0));
+        assertEquals(14, firstIncludeNulls.getInt(1));
+      }
+    }
+  }
+
+  @Test
+  void testGroupByLastIncludeNulls() {
+    try (Table t1 = new Table.TestBuilder()
+            .column(  1,   1,    1,  1,  2,    2,  2,    2)
+            .column(null, 13, null, 12, 14, null, 15, null)
+            .build()) {
+
+      try(Table t2 = t1.groupBy(0)
+              .aggregate(last(1, true));
+          HostColumnVector lastIncludeNulls = t2.getColumn(1).copyToHost()) {
+
+        assertEquals(2, lastIncludeNulls.getRowCount());
+        assertEquals(12, lastIncludeNulls.getInt(0));
+        assert(lastIncludeNulls.isNull(1));
+      }
+    }
+  }
+
   @Test
```

Comment:
We have no tests for the reduce API that returns a scalar.  It might be good to add that in as well.

Reply:
I ended up removing the reduce API since it requires additional features in CuDF